Sources/NIOHTTPCompression/HTTPCompression.swift

```diff
@@ -0,0 +1,130 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOExtrasZlib
+import NIO
+
+public enum HTTPCompression {
+
+    public enum CompressionAlgorithm: String {
+        case gzip
+        case deflate
+    }
+    
+    public enum CompressionError: Error {
+        case uncompressedWritesPending
+        case noDataToWrite
+    }
+
+    struct Compressor {
+        private var stream = z_stream()
+        var isActive = false
+
+        init() { }
+
+        /// Set up the encoder for compressing data according to a specific
+        /// algorithm.
+        mutating func initialize(encoding: CompressionAlgorithm) {
+            assert(!isActive)
+            isActive = true
+            // zlib docs say: The application must initialize zalloc, zfree and opaque before calling the init function.
+            stream.zalloc = nil
+            stream.zfree = nil
+            stream.opaque = nil
+
+            let windowBits: Int32
+            switch encoding {
+            case .deflate:
+                windowBits = 15
+            case .gzip:
+                windowBits = 16 + 15
+            }
+
+            let rc = CNIOExtrasZlib_deflateInit2(&stream, Z_DEFAULT_COMPRESSION, Z_DEFLATED, windowBits, 8, Z_DEFAULT_STRATEGY)
+            precondition(rc == Z_OK, "Unexpected return from zlib init: \(rc)")
+        }
+
+        mutating func compress(inputBuffer: inout ByteBuffer, allocator: ByteBufferAllocator, finalise: Bool) -> ByteBuffer {
+            assert(isActive)
+            let flags = finalise ? Z_FINISH : Z_SYNC_FLUSH
+            guard inputBuffer.readableBytes > 0 || finalise == true else { return allocator.buffer(capacity: 0) }
+            // deflateBound() provides an upper limit on the number of bytes the input can
+            // compress to. We add 5 bytes to handle the fact that Z_SYNC_FLUSH will append
+            // an empty stored block that is 5 bytes long.
```

Comment:
This comment does not match the surrounding code: nothing adds 5.

Reply:
That was from the original `HTTPResponseCompressor` as it was common code I moved it out to a separate file. I guess by moving it I inherited it :-/. I'll go find out what happened here.

Reply:
You got me @adam-fowler , I noticed it was moved and skipped over it :P

Reply:
@weissi skip comment below. Its your code, or at least you copied it into swift-nio-extras

Reply:
I actually think it's _my_ code. At least, the comment reads like I wrote it.

I think the comment was true back when it was written, where we unconditionally used `Z_SYNC_FLUSH`. I believe the 5 bytes comment was the sense in which this can be larger that @glbrntt referred to: specifically, `Z_SYNC_FLUSH` may append an empty block, and that empty block is 5 bytes long.

What I don't know is how the code fell out of step with the comment. I'm not even 100% sure how true the comment is either.

Reply:
Found out why it's 5 bytes. From zlib docs
"If the parameter flush is set to Z_SYNC_FLUSH, all pending output is flushed to the output buffer and the output is aligned on a byte boundary, so that the decompressor can get all input data available so far. (In particular avail_in is zero after the call if enough output space has been provided before the call.) Flushing may degrade compression for some compression algorithms and so it should be used only when necessary. This completes the current deflate block and follows it with an empty stored block that is three bits plus filler bits to the next byte, followed by four bytes (00 00 ff ff). "

Reply:
I'll add in the comment and add the 5 bytes

Reply:
done

Reply:
I don't think adding 5 bytes revolves my comment above though: `deflateBound()` may underestimate the number of bytes required if anything other than `Z_FINISH` or `Z_NO_FLUSH` is used, it's not clear that adding the 5-bytes in case of `Z_SYNC_FLUSH` is sufficient. Given that this code already existed I'm happy if we solve this one separately to this PR.

Reply:
@glbrntt This is the documentation telling you by how much it might over run in the case we use Z_SYNC_FLUSH. It discusses this elsewhere as well.